src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,333 @@
+
+lemma mul_shift_spec (ψ : add_char R R') (a x : R) : mul_shift ψ a x = ψ (of_add (a * x)) := rfl
+
+/-- If `n` is a natural number, then `mul_shift ψ n x = (ψ x) ^ n`. -/
+lemma mul_shift_spec' (ψ : add_char R R') (n : ℕ) (x : R) : mul_shift ψ n x = (ψ x) ^ n :=
+begin
+  rw [mul_shift_spec, ← nsmul_eq_mul, of_add_nsmul, map_pow],
+  refl,
+end
+
+/-- The product of `mul_shift ψ a` and `mul_shift ψ b` is `mul_shift ψ (a + b)`. -/
+lemma mul_shift_mul (ψ : add_char R R') (a b : R) :
+  mul_shift ψ a * mul_shift ψ b = mul_shift ψ (a + b) :=
+begin
+  erw [← monoid_hom.comp_mul ψ],
+  congr,
+  have h : ∀ (f g : R →+ R),
+           f.to_multiplicative * g.to_multiplicative = (f + g).to_multiplicative := λ f g, rfl,
+  rw h,
+  congr,
+  ext,
+  simp only [add_monoid_hom.add_apply, add_monoid_hom.coe_mul_left, add_mul],
+end
+
+/-- `mul_shift ψ 0` is the trivial character. -/
+@[simp]
+lemma mul_shift_zero (ψ : add_char R R') : mul_shift ψ 0 = add_char.trivial :=
+begin
+  ext,
+  simp only [mul_shift, add_char.trivial, monoid_hom.coe_comp, function.comp_app,
+             add_monoid_hom.to_multiplicative_apply_apply, add_monoid_hom.coe_mul_left,
+             zero_mul, of_add_zero, map_one, monoid_hom.coe_mk],
+end
+
+/-- An additive character is *primitive* iff all its multiplicative shifts by nonzero
+elements are nontrivial. -/
+def is_primitive (ψ : add_char R R') : Prop :=
+∀ (a : R), a ≠ 0 → is_nontrivial (mul_shift ψ a)
+
+/-- The map associating to `a : R` the multiplicative shift of `ψ` by `a`
+is injective when `ψ` is primitive. -/
+lemma to_mul_shift_inj_of_is_primitive {ψ : add_char R R'} (hψ : is_primitive ψ) :
+  function.injective ψ.mul_shift :=
+begin
+  intros a b h,
+  have h₁ : mul_shift ψ a * mul_shift ψ (- b) = mul_shift ψ b * mul_shift ψ (- b) :=
+    congr_fun (congr_arg has_mul.mul h) (mul_shift ψ (- b)),
+  simp only [mul_shift_mul, mul_shift_zero, add_right_neg] at h₁,
+  have h₂ := hψ (a + (- b)),
+  rw [h₁, is_nontrivial_iff_ne_trivial, ← sub_eq_add_neg, ne.def add_char.trivial,
+      ne_self_iff_false, imp_false, not_not] at h₂,
+  exact eq_of_sub_eq_zero h₂,
+end
+
+-- `add_comm_group.equiv_direct_sum_zmod_of_fintype`
+-- gives the structure theorem for finite abelian groups.
+-- This could be used to show that the map above is a bijection.
+-- We leave this for a later occasion.
+
+/-- When `R` is a field `F`, then a nontrivial additive character is primitive -/
+lemma is_primitive_of_is_nontrivial {F : Type u} [field F]
+ (ψ : add_char F R') (hψ : is_nontrivial ψ) : is_primitive ψ :=
+begin
+  intros a ha,
+  cases hψ with x h,
+  use (a⁻¹ * x),
+  rwa [mul_shift_spec, mul_inv_cancel_left₀ ha],
+end
+
+/-- Structure for a primitive additive character on a finite ring `R` into a cyclotomic extension
+of a field `R'`. It records which cyclotomic extension it is, the character, and the
+fact that the character is primitive. -/
+@[nolint has_inhabited_instance] -- can't prove that they always exist
```

Comment:
If you don't want to do this then you can just give an example with e.g. R=zmod 2; you only need one instance in total, not one per ring.

Reply:
Is there a good reason for not wanting to do this?

Reply:
I have literally no idea why we lint for this and remember the days when we didn't; I just remember noticing the hack of just making one instance for one example was enough to stop the linter complaining.

Reply:
I guess I'll leave it as is; the `@[nolint ...]` also stops the linter complaining, and I don't like the hack of making a random instance just to satisfy the linter...